sources/packages/api/src/services/education-program-offering/education-program-offering-import-csv.service.ts

```diff
@@ -0,0 +1,227 @@
+      programContext: programsMap[csvModel.sabcProgramCode],
+    }));
+  }
+
+  /**
+   * Get a list of all institution locations indexed by the institution code.
+   * @param institutionId institution to have all the locations retrieved.
+   * @returns list of all institution locations indexed by the institution code.
+   */
+  private async getLocationsMaps(
+    institutionId: number,
+  ): Promise<InstitutionCodeToIdMap> {
+    const locations = await this.institutionLocationService.getLocations(
+      institutionId,
+    );
+    const institutionMap: InstitutionCodeToIdMap = {};
+    locations.forEach((location) => {
+      if (location.institutionCode) {
+        institutionMap[location.institutionCode] = location.id;
+      }
+    });
+    return institutionMap;
+  }
+
+  /**
+   * Get a list of all the education programs presents in the CSV content
+   * indexed by the SABC code.
+   * @param institutionId institution to have all the programs retrieved.
+   * @param csvModels list to extract the SABC codes.
+   * @returns list of all the education programs presents in the CSV content
+   * indexed by the SABC code.
+   */
+  private async getProgramsMaps(
+    institutionId: number,
+    csvModels: OfferingCSVModel[],
+  ): Promise<ProgramCodeToProgramMap> {
+    const distinctProgramCodes = csvModels
+      .filter((model) => !!model.sabcProgramCode) // Remove empty items.
+      .map((model) => model.sabcProgramCode) // Select the program codes.
+      .filter((value, index, codes) => codes.indexOf(value) === index); // Remove duplicated values.
+    const programs = await this.educationProgramService.getProgramsBySABCCodes(
+      institutionId,
+      distinctProgramCodes,
+    );
+    const programMap: ProgramCodeToProgramMap = {};
+    programs.forEach((program) => {
+      if (program.sabcCode) {
+        programMap[program.sabcCode] = program;
+      }
+    });
+    return programMap;
+  }
+
+  /**
+   * Reads a CSV content and generate a CSV object model for each line.
+   * @param csvContent CSV content to generate the models.
+   * @returns CSV object models.
+   */
+  readCSV(csvContent: string): OfferingCSVModel[] {
+    const offeringModels: OfferingCSVModel[] = [];
+    // Remove BOM(Byte order mark), if present.
+    csvContent = removeUTF8BOM(csvContent);
+    const parsedResult = parse(csvContent, {
+      header: true,
+      skipEmptyLines: true,
+    });
+    if (parsedResult.errors.length) {
+      this.logger.error(
+        `The offering CSV parse resulted in some errors. ${JSON.stringify(
+          parsedResult.errors,
+        )}`,
+      );
+      throw new CustomNamedError(
+        "The offering CSV parse resulted in some errors. Please check server errors log for further information.",
+        OFFERING_VALIDATION_CSV_FORMAT_ERROR,
+      );
+    }
+    parsedResult.data.forEach((line) => {
+      const offeringModel = {} as OfferingCSVModel;
+      offeringModels.push(offeringModel);
```

Comment:
correct me if I am wrong. 160 will always push an empty list right(line 158)?

Reply:
Line 160 will push the newly created object (`OfferingCSVModel` not a list just an empty object) into the list of models.
The `offeringModel` is populated right in the next lines.